firebase-database/src/main/java/com/google/firebase/database/core/ServerValues.java

```diff
@@ -32,47 +29,81 @@
     return values;
   }
 
-  public static Object resolveDeferredValue(Object value, Map<String, Object> serverValues) {
-    if (value instanceof Map) {
-      Map mapValue = (Map) value;
-      if (mapValue.containsKey(NAME_SUBKEY_SERVERVALUE)) {
-        String serverValueKey = (String) mapValue.get(NAME_SUBKEY_SERVERVALUE);
-        if (serverValues.containsKey(serverValueKey)) {
-          return serverValues.get(serverValueKey);
-        }
+  public static Object resolveDeferredValue(Object value, Node existing, Map<String, Object> serverValues) {
+    if (!(value instanceof Map)) {
+      return value;
+    }
+    Map mapValue = (Map) value;
+    if (!mapValue.containsKey(NAME_SUBKEY_SERVERVALUE)) {
+      return value;
+    }
+
+    Object op = mapValue.get(NAME_SUBKEY_SERVERVALUE);
+    Object res = null;
+    if (op instanceof String) {
+      res = resolveScalarDeferredValue((String) op, existing, serverValues);
+    } else if (op instanceof Map) {
+      res = resolveComplexDeferredValue((Map) op, existing, serverValues);
+    }
+    if (res == null) {
+      return value;
+    }
+    return res;
+  }
+
+  static Object resolveScalarDeferredValue(String op, Node existing, Map<String, Object> serverValues) {
+      if ("timestamp".equals(op) && serverValues.containsKey(op)) {
+        return serverValues.get(op);
       }
+      return null;
+  }
+
+  static Object resolveComplexDeferredValue(Map<String, Object> op, Node existing, Map<String, Object> serverValues) {
+    // Only supported complex op so far
+    if (!op.containsKey("increment")) {
+      return null;
     }
-    return value;
+
+    Object incrObject = op.get("increment");
+    if (!(incrObject instanceof Number)) {
+      return null;
+    }
+
+    Number increment = (Number)incrObject;
+
+    if (!(existing.isLeafNode() && existing.getValue() instanceof Number)) {
+      return increment;
+    }
+
+    Number existingVal = (Number) existing.getValue();
+    if (longCanHold(increment) && longCanHold(existingVal)) {
```

Comment:
Same comment as in iOS port regarding naming.

Reply:
done